### PR TITLE
CCM-7585 My messages templates landing page testing

### PR DESCRIPTION
**Pull request by m-salaudeen**
## Description

Adding component tests to the manage templates page.

Also includes a couple of functional changes:
- Order the templates by Date created.
- Add time to Date created.

## Context

The component tests were not included in the main functional ticket.

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [x] This PR is a result of pair or mob programming

## Test Evidence

Evidence added to [CCM-7585](https://nhsd-jira.digital.nhs.uk/browse/CCM-7585)

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: infrastructure/terraform/components/app/amplify_domain_association.tf [hunk 1]
```diff
@@ -4,13 +4,13 @@
 #   enable_auto_sub_domain = true
```
Comment (Ian-Hodges) [inline infrastructure/terraform/components/app/amplify_domain_association.tf:4]: Can these comments be deleted? They appear superfluous.

Comment (Ian-Hodges) [inline infrastructure/terraform/components/app/amplify_domain_association.tf:4]: Should these 4 infrastructure changes be here? They seem unrelated to component tests.

Comment (m-salaudeen) [reply infrastructure/terraform/components/app/amplify_domain_association.tf:4]: These are rebased from the main branch, don't know why it came as changes on this PR but they are present in main and is being worked on by someone else so might leave it.

Comment (Ian-Hodges) [reply infrastructure/terraform/components/app/amplify_domain_association.tf:4]: I can see they are meant to be reinstated after a terraform bug has been fixed.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    id: 'valid-email-template',
+    version: 1,
+    name: 'test-template-cat-email',
+    message: 'test example message',
+    subject: 'test example subject',
+    templateType: TemplateType.EMAIL,
+    templateStatus: TemplateStatus.SUBMITTED,
+  }),
+  'text-message': TemplateFactory.create({
+    id: 'valid-sms-template',
```
Comment (chris-elliott-nhsd) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts:25]: there are already templates in the automation tests with these ids - it's risky having multiple tests creating and deleting the same templates, can we use unique ids? just adding '-manage-template-page' to the end of these or something would work

File: infrastructure/terraform/components/app/amplify_domain_association.tf (hunk 1 above)
Comment (Ian-Hodges) [reply infrastructure/terraform/components/app/amplify_domain_association.tf:4]: They are not changes related to this ticket so shouldn't be in this PR.

Comment (Ian-Hodges) [reply infrastructure/terraform/components/app/amplify_domain_association.tf:4]: I think this goes back to my question as to why you are merging this into CCM-5628. It increases the chances of things getting messy.

Comment (Ian-Hodges) [reply infrastructure/terraform/components/app/amplify_domain_association.tf:4]: Merging main into CCM-5628 might clean things up in this PR.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts (hunk 2 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts]: Swapped for uuids.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import { test, expect } from '@playwright/test';
```
[the remark this answers is not in the document]
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts]: Added.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,298 @@
+import { test, expect } from '@playwright/test';
+import { v4 as uuid } from 'uuid';
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:2]: Any reason you can't use `node:crypto` instead? 

```
import { randomUUID } from 'node:crypto'
```

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    page,
+    baseURL,
+  }) => {
+    const manageTemplatesPage = new ManageTemplatesPage(page);
+    await manageTemplatesPage.loadPage();
+
+    expect(page.url()).toContain(`${baseURL}/templates/manage-templates`);
+
+    const email = page.locator(
+      'tr:has-text("email-submitted_manage-templates-page")'
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:119]: Could these be properties on  the `ManageTemplatesPage` class?

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts (hunk 4 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:2]: I wasn't aware of its existence. As we are very unlikely to use this code outside of node.js, I'm happy to switch over.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+  });
+
+  test('should navigate to the manage templates page', async ({
+    page,
+    baseURL,
+  }) => {
+    const manageTemplatesPage = new ManageTemplatesPage(page);
+    await manageTemplatesPage.loadPage();
+    await expect(page).toHaveURL(`${baseURL}/templates/manage-templates`);
+    expect(await manageTemplatesPage.pageHeader.textContent()).toBe(
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:103]: Really minor.

This could be updated to

```
await expect(manageTemplatesPage.pageHeader).toHaveText('Message templates')
```

So the `await` is not nested into the expect. Generally if something bad happens with nested `awaits` the error is swallowed and harder to debug.

Playwright generally has an async matcher for most things.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    const manageTemplatesPage = new ManageTemplatesPage(page);
+    await manageTemplatesPage.loadPage();
+
+    expect(page.url()).toContain(`${baseURL}/templates/manage-templates`);
+
+    const templateRow = page.locator(
+      'tr:has-text("email-submitted_manage-templates-page")'
+    );
+    const templateDeleteLink = templateRow.getByText('Delete', { exact: true });
+    await expect(templateDeleteLink).toBeHidden();
```
Comment (bhansell1) [inline tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:254]: Does this mean the HTML element has a CSS of `display:hidden`?

I'd expect the element to be rendered rather than adding styling to it.

Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:254]: No, it's not rendered at all.
`tobeHidden()` checks if an element is hidden via styling OR not present at all.
I could change it to `await expect(templateDeleteLink).not.toBeInTheDocument();` to specifically check that it's not in the DOM?

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts (hunk 5 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:119]: I've just tried it, but then decided they are probably best where they are. These locators are specific to this test file and the templates loaded into the manage templates page.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts (hunk 4 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts]: Change made.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    await manageTemplatesPage.loadPage();
+    expect(page.url()).toContain(`${baseURL}/templates/manage-templates`);
+    expect(await manageTemplatesPage.pageHeader.textContent()).toBe(
+      'Message templates'
+    );
+    expect(manageTemplatesPage.createTemplateButton).toBeVisible();
+    await manageTemplatesPage.clickCreateTemplateButton();
+    const chooseTemplatePage = await page.waitForSelector('h1');
+    const headerText = await chooseTemplatePage.textContent();
+    expect(headerText).toContain('Choose a template type to create');
```
[the remark this answers is not in the document]
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts]: I agree and it also involves less code. Change made.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts (hunk 6 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts]: Change made.

File: tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts (hunk 7 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-manage-templates-page.component.ts:248]: I'll be honest, it did the same to me. I think Jas made the change originally and I had to check it's implementation.

**Review by bhansell1 — DISMISSED** [review]

**Review by mark-r-bjss — APPROVED** [review]

**Review by gparry333 — DISMISSED** [review]

File: tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts (hunk 2 above)
Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts]: @chris-elliott-nhsd are you happy with the change?

Comment (Ian-Hodges) [reply tests/test-team/template-mgmt-component-tests/template-mgmt-message-template-page.component.ts]: Although, you suggest uuids so I'm assuming you are :-)

**Review by bhansell1 — APPROVED** [review]

**Review by gparry333 — APPROVED** [review]
